cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -107,10 +127,29 @@ public void setAuditPublisher(AuditPublisher auditPublisher) {
    * @throws NotFoundException if the namespace was not found
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists.
    */
-  public Collection<DatasetSpecification> list(Id.Namespace namespace) throws Exception {
+  public Collection<DatasetSpecification> list(final Id.Namespace namespace) throws Exception {
+    Principal principal = SecurityRequestContext.toPrincipal();
     // Throws NamespaceNotFoundException if the namespace does not exist
     ensureNamespaceExists(namespace);
-    return instanceManager.getAll(namespace);
+    Collection<DatasetSpecification> datasets = instanceManager.getAll(namespace);
+    Collection<DatasetId> datasetIds = Collections2.transform(
+      datasets, new Function<DatasetSpecification, DatasetId>() {
+        @Override
+        public DatasetId apply(DatasetSpecification spec) {
+          return namespace.toEntityId().dataset(spec.getName());
+        }
+      });
+    final Set<DatasetId> authorized =
+      authorizationEnforcementService.filter(Sets.newHashSet(datasetIds), principal);
+    Iterable<DatasetSpecification> authorizedDatasets =
```

Comment:
So all you want is return a list of DatasetSpecs, so that DatasetId(namespace, spec.getName()) is authoirzed? It seems that this can be done more easily.